File: python/oauth_client.py
"""OAuth 2.0 Device Authorization Grant client (RFC 8628) — Python connector.

Pairs with ``server/app/api/routers/oauth.py``. Uses only the stdlib (urllib)
so the connector does not gain a runtime dependency on ``requests``.

Public API:
    request_device_code(api_base, client_id, scopes=None) -> dict
    poll_for_token(api_base, client_id, device_code, interval=5, max_wait=900) -> dict
    refresh_token(api_base, client_id, refresh_token) -> dict
    revoke_token(api_base, client_id, token) -> None
    ensure_fresh_token(api_base, client_id, store) -> dict   # high-level

Each token bundle returned matches the dashboard ``/api/oauth/token`` JSON
response, with an additional ``expires_at`` ISO-8601 string we compute
client-side (``now + expires_in``) so the credential store can detect expiry.
"""
from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

import sys
from pathlib import Path

_CONNECTOR_ROOT = Path(__file__).resolve().parent
if str(_CONNECTOR_ROOT) not in sys.path:
    sys.path.insert(0, str(_CONNECTOR_ROOT))
from lib import api_paths as _api_paths


_USER_AGENT = "patcherly-connector-python/1.46"

# Keep in sync with settings_schema connector_local_refresh_retries default.
LOCAL_REFRESH_RETRIES = 3


class RefreshError(RuntimeError):
    """Refresh failure with ``refresh_class`` of transient | auth_death."""

    def __init__(self, message: str, *, refresh_class: str = "transient", http_status: Optional[int] = None):
        super().__init__(message)
        self.refresh_class = refresh_class
        self.http_status = http_status


def classify_refresh_failure(status: Optional[int] = None, body: Any = None, exc: Any = None) -> str:
    """Return ``transient`` or ``auth_death``."""
    if status is not None:
        detail = ""
        if isinstance(body, dict):
            detail = str(body.get("detail") or body.get("error") or body.get("error_description") or "").lower()
        if status in (400, 401) or "invalid_grant" in detail or "invalid_token" in detail or "revoked" in detail:
            return "auth_death"
        if status >= 500 or status in (0, 408, 429):
            return "transient"
        if 400 <= status < 500:
            return "auth_death"
        return "transient"
    msg = str(exc or "").lower()
    if any(x in msg for x in ("timeout", "timed out", "connection", "network", "refused", "reset", "unreachable")):
        return "transient"
    if "invalid_grant" in msg or "invalid_token" in msg:
        return "auth_death"
    import re

    m = re.search(r"http\s+(\d{3})", msg)
    if m:
        return classify_refresh_failure(int(m.group(1)), body)
    return "transient"


def _post_form(api_base: str, path_suffix: str, fields: Dict[str, str]) -> tuple[int, Dict[str, Any]]:
    base = api_base.rstrip("/")
    url = base + path_suffix
    body = urllib.parse.urlencode(fields).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json",
            "User-Agent": _USER_AGENT,
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            raw = resp.read().decode("utf-8") or "{}"
            try:
                parsed = json.loads(raw)
            except json.JSONDecodeError:
                parsed = {"raw": raw}
            return resp.status, parsed
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8", errors="replace") or "{}"
        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError:
            parsed = {"raw": raw}
        return e.code, parsed


def _add_expires_at(bundle: Dict[str, Any]) -> Dict[str, Any]:
    if isinstance(bundle, dict) and isinstance(bundle.get("expires_in"), int):
        ts = (datetime.now(timezone.utc) + timedelta(seconds=bundle["expires_in"])).isoformat()
        bundle = dict(bundle)
        bundle["expires_at"] = ts
    return bundle


def request_device_code(
    api_base: str,
    client_id: str,
    scopes: Optional[List[str]] = None,
) -> Dict[str, Any]:
    fields = {
        "client_id": client_id,
        "scope": " ".join(scopes or ["ingest", "patch", "audit", "files"]),
    }
    status, body = _post_form(api_base, _api_paths.NAMED_PATHS_OAUTH_DEVICE, fields)
    if status != 200:
        raise RuntimeError(f"requestDeviceCode failed (HTTP {status}): {body}")
    return body


def poll_for_token(
    api_base: str,
    client_id: str,
    device_code: str,
    interval: int = 5,
    max_wait_seconds: int = 900,
) -> Dict[str, Any]:
    interval = max(1, int(interval or 5))
    start = time.monotonic()
    while time.monotonic() - start < max_wait_seconds:
        fields = {
            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            "device_code": device_code,
            "client_id": client_id,
        }
        status, body = _post_form(api_base, _api_paths.NAMED_PATHS_OAUTH_TOKEN, fields)
        if status == 200:
            return _add_expires_at(body)
        detail = (body or {}).get("detail", "")
        if detail == "authorization_pending":
            time.sleep(interval)
            continue
        if detail == "slow_down":
            interval += 5
            time.sleep(interval)
            continue
        raise RuntimeError(f"Token exchange failed (HTTP {status}): {body}")
    raise TimeoutError("Device authorization timed out")


def refresh_token(api_base: str, client_id: str, refresh_token_value: str) -> Dict[str, Any]:
    fields = {
        "grant_type": "refresh_token",
        "refresh_token": refresh_token_value,
        "client_id": client_id,
    }
    try:
        status, body = _post_form(api_base, _api_paths.NAMED_PATHS_OAUTH_TOKEN, fields)
    except Exception as e:
        raise RefreshError(
            f"Refresh failed (network): {e}",
            refresh_class="transient",
        ) from e
    if status != 200:
        raise RefreshError(
            f"Refresh failed (HTTP {status}): {body}",
            refresh_class=classify_refresh_failure(status, body),
            http_status=status,
        )
    return _add_expires_at(body)


def revoke_token(
    api_base: str,
    client_id: str,
    token: str,
    *,
    trigger: Optional[str] = None,
) -> None:
    fields: Dict[str, str] = {"token": token, "client_id": client_id}
    if trigger:
        fields["trigger"] = str(trigger)
    _post_form(api_base, _api_paths.NAMED_PATHS_OAUTH_REVOKE, fields)


def signal_disconnect_best_effort(
    api_base: str,
    client_id: str,
    refresh_token_value: Optional[str] = None,
    access_token_value: Optional[str] = None,
    *,
    trigger: str = "auth_failure",
) -> None:
    """Best-effort revoke. Default ``trigger=auth_failure``; logout passes ``logout``."""
    token = refresh_token_value or access_token_value
    if not token:
        return
    try:
        revoke_token(api_base, client_id, token, trigger=trigger)
    except Exception:
        pass


def signal_soft_hold_best_effort(
    api_base: str,
    access_token: Optional[str],
    hmac_secret: Optional[str],
    hmac_kid: Optional[str] = None,
) -> None:
    """Best-effort soft_hold when a bearer+HMAC still works; swallow failures."""
    _signal_reconnect_phase_best_effort(
        api_base, access_token, hmac_secret, hmac_kid, phase="soft_hold", last_error_class="transient"
    )


def signal_reconnect_recovered_best_effort(
    api_base: str,
    access_token: Optional[str],
    hmac_secret: Optional[str],
    hmac_kid: Optional[str] = None,
) -> None:
    """Best-effort recovered ack after a successful nudge/refresh."""
    _signal_reconnect_phase_best_effort(
        api_base, access_token, hmac_secret, hmac_kid, phase="recovered", last_error_class=None
    )


def _signal_reconnect_phase_best_effort(
    api_base: str,
    access_token: Optional[str],
    hmac_secret: Optional[str],
    hmac_kid: Optional[str],
    *,
    phase: str,
    last_error_class: Optional[str],
) -> None:
    if not api_base or not access_token or not hmac_secret:
        return
    try:
        import hashlib
        import hmac as hmac_mod

        path_suffix = _api_paths.NAMED_PATHS_TARGETS_CONNECTOR_RECONNECT_SIGNAL
        payload: Dict[str, Any] = {"phase": phase}
        if last_error_class:
            payload["last_error_class"] = last_error_class
        body = json.dumps(payload).encode("utf-8")
        ts = str(int(time.time()))
        canonical = f"POST\n{path_suffix}\n{ts}\n".encode("utf-8") + body
        sig = hmac_mod.new(hmac_secret.encode("utf-8"), canonical, hashlib.sha256).hexdigest()
        url = api_base.rstrip("/") + path_suffix
        req = urllib.request.Request(
            url,
            data=body,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Authorization": f"Bearer {access_token}",
                "X-Patcherly-Timestamp": ts,
                "X-Patcherly-Signature": sig,
                "User-Agent": _USER_AGENT,
                **({"X-Patcherly-Hmac-Kid": hmac_kid} if hmac_kid else {}),
            },
        )
        with urllib.request.urlopen(req, timeout=15) as resp:
            resp.read()
    except Exception:
        pass


def ensure_fresh_token(api_base: str, client_id: str, store) -> Dict[str, Any]:
    """High-level helper: load creds, refresh if expired, persist, return bundle."""
    creds = store.load()
    if not creds:
        raise RuntimeError(
            "No credentials. Run `patcherly login` to authorize this connector."
        )
    if not store.is_expired(creds):
        return creds
    refresh = creds.get("refresh_token")
    if not refresh:
        signal_disconnect_best_effort(
            api_base,
            client_id,
            None,
            creds.get("access_token") if isinstance(creds.get("access_token"), str) else None,
            trigger="auth_failure",
        )
        raise RuntimeError("Access token expired and no refresh_token available.")

    last_err: Optional[BaseException] = None
    for attempt in range(1, LOCAL_REFRESH_RETRIES + 1):
        try:
            fresh = refresh_token(api_base, client_id, refresh)
            store.save(fresh)
            return fresh
        except RefreshError as e:
            last_err = e
            if e.refresh_class == "auth_death":
                signal_disconnect_best_effort(
                    api_base,
                    client_id,
                    refresh,
                    creds.get("access_token") if isinstance(creds.get("access_token"), str) else None,
                    trigger="auth_failure",
                )
                raise
            if attempt < LOCAL_REFRESH_RETRIES:
                time.sleep(0.5 * attempt)
        except Exception as e:
            last_err = e
            if classify_refresh_failure(exc=e) == "auth_death":
                signal_disconnect_best_effort(
                    api_base,
                    client_id,
                    refresh,
                    creds.get("access_token") if isinstance(creds.get("access_token"), str) else None,
                    trigger="auth_failure",
                )
                raise
            if attempt < LOCAL_REFRESH_RETRIES:
                time.sleep(0.5 * attempt)

    signal_soft_hold_best_effort(
        api_base,
        creds.get("access_token") if isinstance(creds.get("access_token"), str) else None,
        creds.get("hmac_secret") if isinstance(creds.get("hmac_secret"), str) else None,
        creds.get("hmac_secret_id") if isinstance(creds.get("hmac_secret_id"), str) else None,
    )
    if last_err:
        raise last_err
    raise RuntimeError("Refresh failed after transient retries")

File: nodejs/oauth_client.js
/**
 * OAuth 2.0 Device Authorization Grant client (RFC 8628) — Node.js connector.
 *
 * Pairs with the server router at server/app/api/routers/oauth.py. Uses only
 * Node.js built-ins (no axios dep) so the connector stays slim.
 *
 * Public API:
 *   - requestDeviceCode({ apiBase, clientId, scopes })           → { device_code, user_code, expires_in, interval, verification_uri }
 *   - pollForToken({ apiBase, clientId, deviceCode, interval })  → token bundle
 *   - refreshToken({ apiBase, clientId, refreshToken })          → token bundle
 *   - revokeToken({ apiBase, clientId, token, trigger })         → void
 *
 * Token bundle shape (mirrors the dashboard /api/oauth/token response):
 *   {
 *     access_token, refresh_token, expires_in, scope,
 *     hmac_secret, hmac_secret_id, target_id, tenant_id,
 *     // computed:
 *     expires_at: ISO8601,
 *   }
 *
 * The bundle is written verbatim to the local CredentialStore by the CLI.
 *
 * Refresh failure policy (connector auto-reconnect):
 *   - transient (network / 5xx / timeout) → local retries, keep bundle, no revoke
 *   - auth_death (invalid_grant / 401) → revoke with trigger=auth_failure
 *   - logout → revoke with trigger=logout
 */
'use strict';

const apiPaths = require('./lib/api_paths.js');
const { namedPaths } = apiPaths;

const https = require('https');
const http = require('http');
const { URL, URLSearchParams } = require('url');

/** Keep in sync with settings_schema connector_local_refresh_retries default. */
const LOCAL_REFRESH_RETRIES = 3;

function _post(apiBase, pathSuffix, formBody) {
  return new Promise((resolve, reject) => {
    const u = new URL(apiBase.replace(/\/+$/, '') + pathSuffix);
    const body = formBody.toString();
    const lib = u.protocol === 'http:' ? http : https;
    const req = lib.request(
      {
        protocol: u.protocol,
        hostname: u.hostname,
        port: u.port || (u.protocol === 'http:' ? 80 : 443),
        path: u.pathname + u.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': Buffer.byteLength(body),
          Accept: 'application/json',
          'User-Agent': 'patcherly-connector-nodejs/1.46',
        },
      },
      (res) => {
        let buf = '';
        res.setEncoding('utf8');
        res.on('data', (chunk) => (buf += chunk));
        res.on('end', () => {
          let parsed;
          try {
            parsed = buf ? JSON.parse(buf) : {};
          } catch {
            parsed = { raw: buf };
          }
          resolve({ status: res.statusCode, body: parsed });
        });
      },
    );
    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.destroy(new Error('request timeout'));
    });
    req.write(body);
    req.end();
  });
}

function _addExpiresAt(bundle) {
  if (bundle && typeof bundle.expires_in === 'number') {
    const ts = new Date(Date.now() + bundle.expires_in * 1000).toISOString();
    return Object.assign({}, bundle, { expires_at: ts });
  }
  return bundle;
}

/**
 * Classify refresh / token-endpoint failure.
 * @returns {'transient'|'auth_death'}
 */
function classifyRefreshFailure(errOrStatus, body) {
  if (typeof errOrStatus === 'number') {
    const status = errOrStatus;
    const detail = String(
      (body && (body.detail || body.error || body.error_description)) || '',
    ).toLowerCase();
    if (
      status === 400 ||
      status === 401 ||
      detail.includes('invalid_grant') ||
      detail.includes('invalid_token') ||
      detail.includes('revoked')
    ) {
      return 'auth_death';
    }
    if (status >= 500 || status === 408 || status === 429 || status === 0) {
      return 'transient';
    }
    // Other 4xx → treat as auth death (do not soft-hold forever)
    if (status >= 400 && status < 500) return 'auth_death';
    return 'transient';
  }
  const msg = String((errOrStatus && errOrStatus.message) || errOrStatus || '').toLowerCase();
  if (
    msg.includes('timeout') ||
    msg.includes('econnreset') ||
    msg.includes('enotfound') ||
    msg.includes('econnrefused') ||
    msg.includes('network') ||
    msg.includes('socket')
  ) {
    return 'transient';
  }
  const m = msg.match(/http\s+(\d{3})/);
  if (m) return classifyRefreshFailure(parseInt(m[1], 10), body);
  if (msg.includes('invalid_grant') || msg.includes('invalid_token')) return 'auth_death';
  return 'transient';
}

async function requestDeviceCode({ apiBase, clientId, scopes }) {
  const form = new URLSearchParams({
    client_id: clientId,
    scope: (scopes || ['ingest', 'patch', 'audit', 'files']).join(' '),
  });
  const { status, body } = await _post(apiBase, namedPaths.named_paths_oauth_device, form);
  if (status !== 200) {
    throw new Error(
      `requestDeviceCode failed (HTTP ${status}): ${JSON.stringify(body)}`,
    );
  }
  return body;
}

async function pollForToken({
  apiBase,
  clientId,
  deviceCode,
  interval = 5,
  maxWaitSeconds = 900,
}) {
  const start = Date.now();
  let pollInterval = Math.max(1, parseInt(interval, 10) || 5);

  // RFC 8628 §3.5 — poll until 200 or fatal error.
  while ((Date.now() - start) / 1000 < maxWaitSeconds) {
    const form = new URLSearchParams({
      grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
      device_code: deviceCode,
      client_id: clientId,
    });
    const { status, body } = await _post(apiBase, namedPaths.named_paths_oauth_token, form);
    if (status === 200) {
      return _addExpiresAt(body);
    }
    const detail = (body && body.detail) || '';
    if (detail === 'authorization_pending') {
      await new Promise((r) => setTimeout(r, pollInterval * 1000));
      continue;
    }
    if (detail === 'slow_down') {
      pollInterval += 5;
      await new Promise((r) => setTimeout(r, pollInterval * 1000));
      continue;
    }
    throw new Error(
      `Token exchange failed (HTTP ${status}): ${JSON.stringify(body)}`,
    );
  }
  throw new Error('Device authorization timed out');
}

async function refreshToken({ apiBase, clientId, refreshToken }) {
  const form = new URLSearchParams({
    grant_type: 'refresh_token',
    refresh_token: refreshToken,
    client_id: clientId,
  });
  let status;
  let body;
  try {
    ({ status, body } = await _post(apiBase, namedPaths.named_paths_oauth_token, form));
  } catch (e) {
    const err = new Error(`Refresh failed (network): ${e.message || e}`);
    err.refreshClass = 'transient';
    err.cause = e;
    throw err;
  }
  if (status !== 200) {
    const err = new Error(
      `Refresh failed (HTTP ${status}): ${JSON.stringify(body)}`,
    );
    err.refreshClass = classifyRefreshFailure(status, body);
    err.httpStatus = status;
    err.body = body;
    throw err;
  }
  return _addExpiresAt(body);
}

async function revokeToken({ apiBase, clientId, token, trigger }) {
  const form = new URLSearchParams({ token, client_id: clientId });
  if (trigger) form.set('trigger', String(trigger));
  await _post(apiBase, namedPaths.named_paths_oauth_revoke, form);
}

/** Best-effort revoke. Default trigger=auth_failure (hard path). Logout passes trigger=logout. */
async function signalDisconnectBestEffort({
  apiBase,
  clientId,
  refreshToken,
  accessToken,
  trigger = 'auth_failure',
}) {
  const token = refreshToken || accessToken;
  if (!token) return;
  try {
    await revokeToken({ apiBase, clientId, token, trigger });
  } catch {
    // best effort
  }
}

/**
 * Best-effort soft_hold signal when a live access token still works.
 * Swallow failures (expired bearer → server sweep owns silence soft_hold).
 */
async function signalSoftHoldBestEffort({ apiBase, accessToken, hmacSecret, hmacKid }) {
  return signalReconnectPhaseBestEffort({
    apiBase,
    accessToken,
    hmacSecret,
    hmacKid,
    phase: 'soft_hold',
    lastErrorClass: 'transient',
  });
}

/** Best-effort recovered ack after a successful server reconnect nudge. */
async function signalReconnectRecoveredBestEffort({ apiBase, accessToken, hmacSecret, hmacKid }) {
  return signalReconnectPhaseBestEffort({
    apiBase,
    accessToken,
    hmacSecret,
    hmacKid,
    phase: 'recovered',
  });
}

async function signalReconnectPhaseBestEffort({
  apiBase,
  accessToken,
  hmacSecret,
  hmacKid,
  phase,
  lastErrorClass,
}) {
  if (!apiBase || !accessToken || !hmacSecret) return;
  try {
    const crypto = require('crypto');
    const pathSuffix = namedPaths.named_paths_targets_connector_reconnect_signal;
    const payload = { phase };
    if (lastErrorClass) payload.last_error_class = lastErrorClass;
    const body = JSON.stringify(payload);
    const ts = Math.floor(Date.now() / 1000).toString();
    const method = 'POST';
    const canonical = [method, pathSuffix, ts, body].join('\n');
    const sig = crypto.createHmac('sha256', hmacSecret).update(canonical).digest('hex');
    const u = new URL(apiBase.replace(/\/+$/, '') + pathSuffix);
    const lib = u.protocol === 'http:' ? http : https;
    await new Promise((resolve, reject) => {
      const headers = {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
        Accept: 'application/json',
        Authorization: `Bearer ${accessToken}`,
        'X-Patcherly-Timestamp': ts,
        'X-Patcherly-Signature': sig,
        'User-Agent': 'patcherly-connector-nodejs/1.46',
      };
      if (hmacKid) headers['X-Patcherly-Hmac-Kid'] = hmacKid;
      const req = lib.request(
        {
          protocol: u.protocol,
          hostname: u.hostname,
          port: u.port || (u.protocol === 'http:' ? 80 : 443),
          path: u.pathname + u.search,
          method: 'POST',
          headers,
        },
        (res) => {
          res.resume();
          res.on('end', resolve);
        },
      );
      req.on('error', reject);
      req.write(body);
      req.end();
    });
  } catch {
    // best effort
  }
}

/** High-level convenience: returns a fresh access token, refreshing if needed. */
async function ensureFreshToken({ apiBase, clientId, store }) {
  const creds = store.load();
  if (!creds) {
    throw new Error(
      'No credentials. Run `patcherly login` to authorize this connector.',
    );
  }
  if (!store.isExpired(creds)) return creds;
  if (!creds.refresh_token) {
    await signalDisconnectBestEffort({
      apiBase,
      clientId,
      refreshToken: null,
      accessToken: creds.access_token,
      trigger: 'auth_failure',
    });
    throw new Error('Access token expired and no refresh_token available.');
  }

  let lastErr = null;
  for (let attempt = 1; attempt <= LOCAL_REFRESH_RETRIES; attempt++) {
    try {
      const fresh = await refreshToken({
        apiBase,
        clientId,
        refreshToken: creds.refresh_token,
      });
      store.save(fresh);
      return fresh;
    } catch (e) {
      lastErr = e;
      const klass = e.refreshClass || classifyRefreshFailure(e);
      if (klass === 'auth_death') {
        await signalDisconnectBestEffort({
          apiBase,
          clientId,
          refreshToken: creds.refresh_token,
          accessToken: creds.access_token,
          trigger: 'auth_failure',
        });
        throw e;
      }
      if (attempt < LOCAL_REFRESH_RETRIES) {
        await new Promise((r) => setTimeout(r, 500 * attempt));
      }
    }
  }

  // Exhausted local transient retries — keep bundle, soft-hold signal best-effort.
  await signalSoftHoldBestEffort({
    apiBase,
    accessToken: creds.access_token,
    hmacSecret: creds.hmac_secret,
    hmacKid: creds.hmac_secret_id,
  });
  throw lastErr || new Error('Refresh failed after transient retries');
}

module.exports = {
  LOCAL_REFRESH_RETRIES,
  classifyRefreshFailure,
  requestDeviceCode,
  pollForToken,
  refreshToken,
  revokeToken,
  signalDisconnectBestEffort,
  signalSoftHoldBestEffort,
  signalReconnectRecoveredBestEffort,
  ensureFreshToken,
};
